From a519a64bdc979e9d9b09d715a94b692fa0485f38 Mon Sep 17 00:00:00 2001
From: Isaac Bosompem <mrisaacb@google.com>
Date: Wed, 18 Aug 2021 13:43:52 -0400
Subject: [PATCH] glamor: Fix xRGB->ARGB composite in certain cases

Currently an ARGB window under Xwayland rootless mode
will initially be filled with opaque black pixels as
a result of the screen pixmap (a "virtual" 0x0 pixmap)
being composited onto the window (see compNewPixmap).

This was due to convert_gradient_picture expanding this
pixmap to the same size as the composite region. Then,
subsequently performing an xRGB->ARGB translation.

This removes the expansion code for this case and
adds code to check for this case and replace the
incoming 0x0 pixmap with a solid color.

Signed-off By: Isaac Bosompem <mrisaacb@google.com>
---
 glamor/glamor_render.c | 40 ++++++++++++++++++++++++++--------------
 1 file changed, 26 insertions(+), 14 deletions(-)

diff --git a/glamor/glamor_render.c b/glamor/glamor_render.c
index 2af65bf93..00ec03c71 100644
--- a/glamor/glamor_render.c
+++ b/glamor/glamor_render.c
@@ -800,6 +800,20 @@ glamor_render_format_is_supported(PicturePtr picture)
     }
 }
 
+/**
+* Returns whether a picture w/ a drawable can be replaced with a solid block
+* of pixels.
+*
+* This will be true if either of the dimensions are equal to 0.
+*/
+static Bool
+glamor_composite_can_replace_with_solid(PicturePtr picture)
+{
+    PixmapPtr pixmap = glamor_get_drawable_pixmap(picture->pDrawable);
+
+    return (!pixmap->drawable.width || !pixmap->drawable.height);
+}
+
 static Bool
 glamor_composite_choose_shader(CARD8 op,
                                PicturePtr source,
@@ -855,6 +869,13 @@ glamor_composite_choose_shader(CARD8 op,
         else
             goto fail;
     }
+    else if (glamor_composite_can_replace_with_solid(source)) {
+        key.source = SHADER_SOURCE_SOLID;
+        source_solid_color[0] = 0.0;
+        source_solid_color[1] = 0.0;
+        source_solid_color[2] = 0.0;
+        source_solid_color[3] = 0.0;
+    }
     else {
         if (PICT_FORMAT_A(source->format))
             key.source = SHADER_SOURCE_TEXTURE_ALPHA;
@@ -873,6 +894,9 @@ glamor_composite_choose_shader(CARD8 op,
             else
                 goto fail;
         }
+        else if (glamor_composite_can_replace_with_solid(mask)) {
+            key.mask = SHADER_MASK_NONE;
+        }
         else {
             if (PICT_FORMAT_A(mask->format))
                 key.mask = SHADER_MASK_TEXTURE_ALPHA;
@@ -1470,13 +1494,7 @@ glamor_composite_clipped_region(CARD8 op,
     }
 
     /* XXX is it possible source mask have non-zero drawable.x/y? */
-    if (source
-        && ((!source->pDrawable
-             && (source->pSourcePict->type != SourcePictTypeSolidFill))
-            || (source->pDrawable
-                && !GLAMOR_PIXMAP_PRIV_HAS_FBO(source_pixmap_priv)
-                && (source_pixmap->drawable.width != width
-                    || source_pixmap->drawable.height != height)))) {
+    if (source && source->pSourcePict && source->pSourcePict->type != SourcePictTypeSolidFill) {
         temp_src =
             glamor_convert_gradient_picture(screen, source,
                                             extent->x1 + x_source - x_dest - dest->pDrawable->x,
@@ -1492,13 +1510,7 @@ glamor_composite_clipped_region(CARD8 op,
         y_temp_src = -extent->y1 + y_dest + dest->pDrawable->y;
     }
 
-    if (mask
-        &&
-        ((!mask->pDrawable
-          && (mask->pSourcePict->type != SourcePictTypeSolidFill))
-         || (mask->pDrawable && !GLAMOR_PIXMAP_PRIV_HAS_FBO(mask_pixmap_priv)
-             && (mask_pixmap->drawable.width != width
-                 || mask_pixmap->drawable.height != height)))) {
+    if (mask && mask->pSourcePict && mask->pSourcePict->type != SourcePictTypeSolidFill) {
         /* XXX if mask->pDrawable is the same as source->pDrawable, we have an opportunity
          * to do reduce one conversion. */
         temp_mask =
-- 
GitLab

